src/target/target_flash.c

```diff
@@ -143,7 +143,10 @@ bool target_flash_erase(target *t, target_addr_t addr, size_t len)
 		if (!flash_prepare(f))
 			return false;
 
-		ret &= f->erase(f, local_start_addr, f->blocksize);
+		if (!f->erase(f, local_start_addr, f->blocksize)) {
+			DEBUG_WARN("Erase failed at %" PRIx32 "\n", local_start_addr);
+			return false;
```

Comment:
There is major problems here - calling `flash_done()` prior to exiting from this routine is not optional. It must be done to create the correct semantics for a variety of devices and perform buffer allocation cleanup that is required to avoid heap exhaustion.

The other reason we need `flash_done()` called is that in several cases, the correct functioning of the device hinges on it as entering Flash mode inhibits operations like reading the Flash mapping region. As mentioned, not an optional call.

There are two solutions we see here - the first is to insert a call to `flash_done()` just prior to this return (ignoring the return value) - which is not.. great.. the second is to have the old line, then an if check for ret - if ret is false, issue the debug warn and then break, moving the flash_done at the bottom of the loop one line down outside the loop. We favour the latter option.